protocol/contracts/v2/strategies/SkyFarmBaseStrategy.sol

```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * @notice The TRV is able to withdraw on demand in order to fund other strategies which 
+     * wish to borrow from the TRV.
+     * @dev It may withdraw less than requested if there isn't enough balance in the `usdsVaultToken`.
+     */
+    function trvWithdraw(uint256 requestedAmount) external override returns (uint256) {
+        address _trvAddr = address(treasuryReservesVault);
+        if (msg.sender != _trvAddr) revert OnlyTreasuryReserveVault(msg.sender);
+        if (requestedAmount == 0) revert CommonEventsAndErrors.ExpectedNonZero();       
+
+        // Cap at the max amount which can be withdrawn.
+        uint256 availableToWithdraw = usdsVaultToken.maxWithdraw(address(this));
+        if (requestedAmount > availableToWithdraw) {
+            requestedAmount = availableToWithdraw;
+        }
+
+        _daiWithdrawal(requestedAmount);
+        daiToken.safeTransfer(_trvAddr, requestedAmount);
+        return requestedAmount;
+    }
+
+    /**
+     * @notice A hook where strategies can optionally update approvals when the trv is updated
+     */
+    function _updateTrvApprovals(address oldTrv, address newTrv) internal override {
+        _setMaxAllowance(daiToken, oldTrv, newTrv);
+    }
+
+    function _daiDeposit(uint256 daiAmount) internal {
+        // Convert to USDS
+        daiToUsds.daiToUsds(address(this), daiAmount);
+
+        // Deposit into the USDS savings vault
+        usdsVaultToken.deposit(daiAmount, address(this));
+        emit DaiDeposited(daiAmount);
+    }
+
+    function _daiWithdrawal(uint256 daiAmount) internal {
+        // Withdraw from the USDS savings vault
+        usdsVaultToken.withdraw(daiAmount, address(this), address(this));
+
+        // Convert to DAI
+        daiToUsds.usdsToDai(address(this), daiAmount);
+        emit DaiWithdrawn(daiAmount);
+    }
+
+    function _vaultRedemption(uint256 vaultShares) internal returns (uint256 daiAmount) {
+        // Redeem from the USDS savings vault
+        daiAmount = usdsVaultToken.redeem(vaultShares, address(this), address(this));
+
+        // Convert to DAI
+        daiToUsds.usdsToDai(address(this), daiAmount);
+        emit DaiWithdrawn(daiAmount);
+    }
+
+    /**
+     * @notice The strategy executor can shutdown this strategy, only after Governance has 
+     * marked the strategy as `isShuttingDown` in the TRV.
+     * This should handle all liquidations and send all funds back to the TRV, and will then call `TRV.shutdown()`
+     * to apply the shutdown.
+     * @dev Each strategy may require a different set of params to do the shutdown. It can abi encode/decode
+     * that data off chain, or by first calling populateShutdownData()
+     * Shutdown data isn't required for the `usdsVaultToken` automated shutdown.
+     */
+    function _doShutdown(bytes calldata /*data*/) internal override {
+        // Withdraw all
+        uint256 daiAmount = _vaultRedemption(usdsVaultToken.maxRedeem(address(this)));
```

Comment:
It bothers me a bit that this `_vaultRedemption()` is only triggered when shutting down with `automatedShutdown()` but not if `ElevatedAccess` calls directly `TRV.shutdown()`. 

In that case, the redemption cannot be performed unless you enable the strategy again, and then do a proper shutdown with `automatedShutdown()` , so it is not the end of the world. But it bothers me so I had to point it out